### DISCUSSION TITLE
Migration and postgresql schemas

### DISCUSSION THREAD

**Discussion by pphunor72** [mikro-orm/mikro-orm#1886] [General]
Hello,

I try to create/run migrations for my schema based multi tenant application, and I cannot figure out the right solution. My problem is that I cannot add schema name to migrations.tableName config option, and schema name specified at driverOptions.schema has no effect here.

Both MigrationStorage.ensureTable (migrations) and DatabaseSchema.create (knex/schema) ignores schema names returned by getListTablesSQL, thus the comparision with table name (that comes from migrations.tableName) does not give the expected result. It is worth noting that getListTablesSQL returns tables for all schemas regardless of connection and driverOptions.schema, is this good?

Schema name is not set at my entity definitions, I used to set it runtime. I still do not know if this will be a problem or not after overcoming this migration table issue.
However the ideal solution would be that if a schema is specified in a connection setup, then table names and any other items are interpreted in the context of that schema, except where another schema is explicitly specified.

Thanks for your effort creating and maintaing this project.

P.S. I use mikro-orm version 4.5.5.

**Comment by B4nan** [comment]
Multitenancy is not really supported, at least I would not call that something supported - there are bits a pieces but not as a whole. It is something I never really tried myself so can't help much. There are two PRs that are trying to add some kind of support for that (#1001 and #1424), both are WIP and tbh I am still not sure what way to go about it. Will be happy to hear concrete suggestions.

For migrations you could probably go for the programatic approach and handle what ever you need yourself (e.g. you could instantiate the ORM multiple times with different schema name/search path and use the migrator from it).

> driverOptions.schema has no effect here

That sounds wrong, `driverOptions` are for the whole knex config, and there is no top level `schema` option. It should be `driverOptions.connection.schema` instead.

> It is worth noting that getListTablesSQL returns tables for all schemas 

I guess you should be able to control this via search path. The thing is that if we limit it to one schema only, diffing of entities that are in a different schema could not work. It would have to be an array of available schemas, not a single one. 

> regardless of connection and driverOptions.schema

Don't expect much effect on internals (mainly those schema related queries) when you change knex configuration (via `driverOptions`). Most of the queries are not done via knex's qb, but a raw sql instead.

In v5 you can configure the schema name directly via MikroORM configuration. 

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/Configuration.ts#L292

> However the ideal solution would be that if a schema is specified in a connection setup, then table names and any other items are interpreted in the context of that schema, except where another schema is explicitly specified.

Makes sense, we can make it work that way in v5. Currenlty you need to provide the schema name explicitly via `FindOptions` (as in v4 there is no way to set the schema on a connection level).

**Reply by pphunor72** [reply]
> > It is worth noting that getListTablesSQL returns tables for all schemas
> 
> I guess you should be able to control this via search path. The thing is that if we limit it to one schema only, diffing of entities that are in a different schema could not work. It would have to be an array of available schemas, not a single one.

Unfortunately adjusting search path won't help here, getListTablesSQL uses information_schema directly, and always returns all tables from all schemas.

> 
> > regardless of connection and driverOptions.schema
> 
> Don't expect much effect on internals (mainly those schema related queries) when you change knex configuration (via `driverOptions`). Most of the queries are not done via knex's qb, but a raw sql instead.
> 
> In v5 you can configure the schema name directly via MikroORM configuration.
> 
> https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/Configuration.ts#L292
> 

Do you have some schedule for version 5? Thanks.

> > However the ideal solution would be that if a schema is specified in a connection setup, then table names and any other items are interpreted in the context of that schema, except where another schema is explicitly specified.
> 
> Makes sense, we can make it work that way in v5. Currenlty you need to provide the schema name explicitly via `FindOptions` (as in v4 there is no way to set the schema on a connection level).

**Reply by B4nan** [reply]
Give it a few months, it's hard to estimate during summer.

**Reply by pphunor72** [reply]
Finally it seems I figured out this tenant schema migration process. The key findings are:
- use migrator programmatically
- set the search path to desired schema.
- use unique migration table names in database to work around getListTablesSQL issue. My solution appends the schema name to 'migrations_' string, so there are tables called 'migrations_tenant_one', 'migrations_tenant_two' and so on, each of them placed in the corresponding schema.
